Stripe/STPPaymentContext.m

```diff
@@ -568,7 +544,116 @@ - (void)requestSynchronousSourcePayment:(id<STPSourceProtocol>)source {
     }];
 }
 
+- (void)requestCardSourcePayment:(STPSource *)source {
+    if (self.state != STPPaymentContextStateNone) {
+        return;
+    }
+
+    if (source.type != STPSourceTypeCard) {
+        return;
+    }
+
+    STPSourceCard3DSecureStatus threeDSecureStatus;
+
+    if (source.cardDetails != nil) {
+        threeDSecureStatus = source.cardDetails.threeDSecure;
+    }
+    else {
+        threeDSecureStatus = STPSourceCard3DSecureStatusNotSupported;
+    }
+
+    STPThreeDSecureSupportType threeDSecureMethodToUse = self.configuration.threeDSecureSupportTypeBlock();
+
+    switch (threeDSecureMethodToUse) {
+        case STPThreeDSecureSupportTypeStatic: {
+            switch (threeDSecureStatus) {
+                case STPSourceCard3DSecureStatusRequired:
+                case STPSourceCard3DSecureStatusOptional:
+                    [self requestThreeDSSourceCreationAndPayment:source];
+                    break;
+                case STPSourceCard3DSecureStatusNotSupported:
+                case STPSourceCard3DSecureStatusUnknown:
+                    [self requestSynchronousSourcePayment:source];
+                    break;
+            }
+        }
+            break;
+        case STPThreeDSecureSupportTypeDynamic: {
+            switch (threeDSecureStatus) {
+                case STPSourceCard3DSecureStatusRequired:
+                    [self requestThreeDSSourceCreationAndPayment:source];
+                    break;
+                case STPSourceCard3DSecureStatusOptional:
+                    [self requestPrecheckAndTakeRequiredActions:source];
+                    break;
+                case STPSourceCard3DSecureStatusNotSupported:
+                case STPSourceCard3DSecureStatusUnknown:
+                    [self requestSynchronousSourcePayment:source];
+                    break;
+            }
+        }
+            break;
+        case STPThreeDSecureSupportTypeDisabled:
+            [self requestSynchronousSourcePayment:source];
+            break;
+    }
+}
+
+- (void)requestPrecheckAndTakeRequiredActions:(STPSource *)source {
+    if (self.state != STPPaymentContextStateNone) {
+        return;
+    }
+
+    self.state = STPPaymentContextStateRequestingPayment;
+
+    STPSourcePrecheckParams *precheckParams = [STPSourcePrecheckParams new];
+    precheckParams.sourceID = source.stripeID;
+    precheckParams.paymentAmount = @(self.paymentAmount);
+    precheckParams.paymentCurrency = self.paymentCurrency;
+    // TODO: method for merchant to add metadata/additional fields
```

Comment:
Meta data for precheck is not the same as source creation metadata (one is metadata you would attach to charge, the other is for attaching to source). Adding new properties for ithem onto additional source info could work maybe.